src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs

```diff
@@ -0,0 +1,136 @@
+    /// dispose of its corresponding JVM thread. All methods are thread-safe.
+    /// </summary>
+    internal class JvmThreadPoolGC : IDisposable
+    {
+        private readonly IJvmBridge _jvmBridge;
+        private readonly TimeSpan _threadGCInterval;
+        private readonly ConcurrentDictionary<int, Thread> _activeThreads;
+
+        private readonly object _activeThreadGCTimerLock;
+        private Timer _activeThreadGCTimer;
+
+        /// <summary>
+        /// Construct the JvmThreadPoolGC.
+        /// </summary>
+        /// <param name="jvmBridge">The JvmBridge used to call JVM methods.</param>
+        /// <param name="threadGCInterval">The interval to GC finished threads.</param>
+        public JvmThreadPoolGC(IJvmBridge jvmBridge, TimeSpan threadGCInterval)
+        {
+            _jvmBridge = jvmBridge;
+            _threadGCInterval = threadGCInterval;
+            _activeThreads = new ConcurrentDictionary<int, Thread>();
+
+            _activeThreadGCTimerLock = new object();
+            _activeThreadGCTimer = null;
+        }
+
+        /// <summary>
+        /// Dispose of the GC timer and run a final round of thread GC.
+        /// </summary>
+        public void Dispose()
+        {
+            lock (_activeThreadGCTimerLock)
+            {
+                if (_activeThreadGCTimer != null)
+                {
+                    _activeThreadGCTimer.Dispose();
+                    _activeThreadGCTimer = null;
+                }
+            }
+
+            GCThreads();
+        }
+
+        /// <summary>
+        /// Try to start monitoring a thread.
+        /// </summary>
+        /// <param name="thread">The thread to add.</param>
+        /// <returns>True if success, false if already added.</returns>
+        public bool TryAddThread(Thread thread)
+        {
+            bool returnValue = _activeThreads.TryAdd(thread.ManagedThreadId, thread);
+
+            // Initialize the GC timer if necessary.
+            if (_activeThreadGCTimer == null)
+            {
+                lock (_activeThreadGCTimerLock)
+                {
+                    if (_activeThreadGCTimer == null && _activeThreads.Count > 0)
+                    {
+                        _activeThreadGCTimer = new Timer(
+                            (state) => GCThreads(),
+                            null,
+                            _threadGCInterval,
+                            _threadGCInterval);
+                    }
+                }
+            }
+
+            return returnValue;
+        }
+
+        /// <summary>
+        /// Try to remove a thread from the pool. If the removal is successful, then the
+        /// corresponding JVM thread will also be disposed.
+        /// </summary>
+        /// <param name="threadId">The ID of the thread to remove.</param>
+        /// <returns>True if success, false if the thread cannot be found.</returns>
+        private bool TryRemoveAndDisposeThread(int threadId)
```

Comment:
nit: How about `TryDisposeJvmThread`? This is doing `TryRemoveDotnetThreadAndDisposeJvmThread` but it's mouthful 😄.

Reply:
Changed.